transmogrifier/sources/transformer.py

```diff
@@ -362,70 +328,98 @@ def record_is_deleted(cls, source_record: dict[str, JSON] | Tag) -> bool:
             source_record: A single source record.
         """
 
-    def get_optional_fields(self, _source_record: dict[str, JSON] | Tag) -> dict | None:
+    @final
+    def get_optional_field_methods(self) -> Iterator[tuple[str, Callable]]:
         """
-        Retrieve optional TIMDEX fields from a source record.
+        Return optional TIMDEX field names and corresponding methods.
 
-        May be overridden by source subclasses.
+        May not be overridden.
+        """
+        for field_name in [
+            field_name
+            for field_name in timdex.TimdexRecord.get_optional_field_names()
+            if hasattr(self, f"get_{field_name}")
+        ]:
+            yield field_name, getattr(self, f"get_{field_name}")
 
-        Args:
-            source_record: A single source record.
+    @final
+    def generate_derived_fields(
+        self, timdex_record: timdex.TimdexRecord
+    ) -> timdex.TimdexRecord:
         """
-        return {}
+        Generate field values based on existing values in TIMDEX record.
+
+        May not be overridden.
+        """
+        self.create_dates_and_locations_from_publishers(timdex_record)
+        self.create_locations_from_spatial_subjects(timdex_record)
+
+        if timdex_record.citation is None:
+            timdex_record.citation = generate_citation(timdex_record)
+        if timdex_record.content_type is None:
+            timdex_record.content_type = ["Not specified"]
+
+        return timdex_record
```

Comment:
In our discussions, we've referred to these steps are "post transform"/"post processing" and am wondering if the new function can be renamed to include these phrases? 🤔

Reply:
I forgot to add a comment about this: as I thought more about this, I think the derived fields should be considered part of the `transform`. The transformed record isn't really "done" until we add the derived final so that transform/post-transform split doesn't feel as important as when we first discussed it. 

I think it's important to keep it as a separate method from the required/optional fields loop but that method still belongs in `transform` in my mind. Regardless of my views though, I also want to hear both of your thoughts on this!

Reply:
Thank you both for explaining! It helps to view it this way, and I agree, `generate_derived_fields` feels part of the `transform` compared to performing validation or metrics.